File: src/csharp/Foundgine.Core/Semantic/Planning/AggregateRelationshipFilterPushdownRule.cs
using Foundgine.Core.Semantic.Query;

namespace Foundgine.Core.Semantic.Planning;

/// <summary>
/// Pushes an existential relationship predicate into a COUNT aggregate.
///
/// The transformation is only applied to the proven-equivalent shape:
/// COUNT(R) > 0 (or >= 1 / != 0) AND SOME(R, P)
/// becomes COUNT(R WHERE P) > 0.
///
/// This keeps the semantic meaning intact while allowing providers to evaluate
/// the child predicate inside the aggregate subquery instead of maintaining a
/// separate relationship-existence check.
/// </summary>
public sealed class AggregateRelationshipFilterPushdownRule : IPlanRewriteRule
{
    public string Name => "aggregate.relationship-filter.pushdown";

    public IReadOnlyList<string> Preconditions =>
    [
        "plan contains an AND filter",
        "AND contains an eligible COUNT existence predicate",
        "AND contains a SOME relationship predicate for the same relationship",
        "COUNT aggregate does not already contain a predicate"
    ];

    public IReadOnlyList<string> SecurityObligations =>
    [
        "authorization.required",
        "authorization.runtime",
        "visibility.relationship",
        "planning.cache-context-isolation"
    ];

    public double CostImpact => 1.25d;
    public double BenefitEstimate => 4.0d;
    public bool IsIdempotent => true;
    public int Priority => 35;

    public IReadOnlyList<string> MustRunBefore =>
    [
        "aggregate.cardinality.short-circuit"
    ];

    public bool CanApply(SemanticPlan plan)
    {
        ArgumentNullException.ThrowIfNull(plan);
        return ContainsEligible(plan.Root);
    }

    public SemanticPlan Apply(SemanticPlan plan)
    {
        ArgumentNullException.ThrowIfNull(plan);
        if (!CanApply(plan))
            return plan;

        var changed = false;
        var root = RewriteNode(plan.Root, ref changed);
        return changed ? new SemanticPlan(root, plan.RequiredSecurityInvariants, plan.AuthorizationBinding) : plan;
    }

    private static SemanticPlanNode RewriteNode(SemanticPlanNode node, ref bool changed)
    {
        var options = node.QueryOptions;
        var filterChanged = false;
        if (options?.Filter is not null)
        {
            var originalFilter = options.Filter;
            var rewritten = RewriteFilter(originalFilter, ref changed);
            filterChanged = !ReferenceEquals(rewritten, originalFilter);
            if (filterChanged)
                options = options with { Filter = rewritten };
        }

        var children = new SemanticPlanNode[node.Children.Count];
        var childrenChanged = false;
        for (var i = 0; i < node.Children.Count; i++)
        {
            children[i] = RewriteNode(node.Children[i], ref changed);
            childrenChanged |= !ReferenceEquals(children[i], node.Children[i]);
        }

        if (childrenChanged)
            changed = true;

        if (filterChanged)
            changed = true;

        return changed && (options is not null || childrenChanged)
            ? node with
            {
                QueryOptions = options,
                Children = children,
                // The pushdown changes the aggregate's semantic shape. Any
                // previously attached cardinality hint was derived from the
                // old bare COUNT and is therefore no longer valid. The
                // cardinality rule may re-derive a hint later if the new shape
                // proves eligible.
                AggregateExecutionStrategy = filterChanged
                    ? AggregateExecutionStrategy.Default
                    : node.AggregateExecutionStrategy
            }
            : node;
    }

    private static SemanticFilterExpression RewriteFilter(
        SemanticFilterExpression filter,
        ref bool changed)
    {
        switch (filter)
        {
            case SemanticAndFilter and:
            {
                var expressions = and.Expressions.ToArray();

                for (var i = 0; i < expressions.Length; i++)
                {
                    if (expressions[i] is not SemanticAggregateFilter aggregate ||
                        !IsEligibleCountExists(aggregate) ||
                        aggregate.Predicate is not null)
                    {
                        continue;
                    }

                    for (var j = 0; j < expressions.Length; j++)
                    {
                        if (i == j ||
                            expressions[j] is not SemanticRelationshipFilter relationship ||
                            relationship.Quantifier != SemanticRelationshipQuantifier.Some ||
                            relationship.Relationship != aggregate.Relationship)
                        {
                            continue;
                        }

                        var rewrittenAggregate = aggregate with
                        {
                            Predicate = relationship.Predicate
                        };

                        var remaining = new List<SemanticFilterExpression>(
                            expressions.Length - 1);

                        for (var k = 0; k < expressions.Length; k++)
                        {
                            if (k == i)
                            {
                                remaining.Add(rewrittenAggregate);
                            }
                            else if (k != j)
                            {
                                remaining.Add(expressions[k]);
                            }
                        }

                        changed = true;

                        return remaining.Count switch
                        {
                            0 => throw new InvalidOperationException(
                                "Aggregate pushdown produced an empty AND expression."),

                            1 => remaining[0],

                            _ => new SemanticAndFilter(remaining)
                        };
                    }
                }

                var nested = new SemanticFilterExpression[expressions.Length];
                var nodeChanged = false;

                for (var i = 0; i < expressions.Length; i++)
                {
                    var rewritten = RewriteFilter(
                        expressions[i],
                        ref changed);

                    nested[i] = rewritten;

                    if (!ReferenceEquals(rewritten, expressions[i]))
                        nodeChanged = true;
                }

                if (!nodeChanged)
                    return filter;

                changed = true;
                return new SemanticAndFilter(nested);
            }

            case SemanticOrFilter or:
            {
                var expressions = or.Expressions.ToArray();
                var nested = new SemanticFilterExpression[expressions.Length];
                var nodeChanged = false;

                for (var i = 0; i < expressions.Length; i++)
                {
                    var rewritten = RewriteFilter(
                        expressions[i],
                        ref changed);

                    nested[i] = rewritten;

                    if (!ReferenceEquals(rewritten, expressions[i]))
                        nodeChanged = true;
                }

                if (!nodeChanged)
                    return filter;

                changed = true;
                return new SemanticOrFilter(nested);
            }

            case SemanticRelationshipFilter relationship:
            {
                var predicate = RewriteFilter(
                    relationship.Predicate,
                    ref changed);

                if (ReferenceEquals(predicate, relationship.Predicate))
                    return filter;

                return relationship with
                {
                    Predicate = predicate
                };
            }

            default:
                return filter;
        }
    }

    private static bool ContainsEligible(SemanticPlanNode node)
    {
        if (node.QueryOptions?.Filter is not null && ContainsEligible(node.QueryOptions.Filter))
            return true;
        return node.Children.Any(ContainsEligible);
    }

    private static bool ContainsEligible(SemanticFilterExpression filter) =>
        filter switch
        {
            SemanticAndFilter and =>
                and.Expressions.OfType<SemanticAggregateFilter>().Any(IsEligibleCountExists) &&
                and.Expressions.OfType<SemanticRelationshipFilter>()
                    .Any(r => r.Quantifier == SemanticRelationshipQuantifier.Some),
            SemanticRelationshipFilter relationship => ContainsEligible(relationship.Predicate),
            SemanticOrFilter or => or.Expressions.Any(ContainsEligible),
            _ => false
        };

    private static bool IsEligibleCountExists(SemanticAggregateFilter aggregate)
    {
        if (aggregate.Aggregate != SemanticFilterAggregate.Count || aggregate.Field is not null ||
            aggregate.Predicate is not null)
            return false;

        return AggregateExecutionStrategyResolver.Resolve(aggregate.Operator, aggregate.Value) ==
               AggregateExecutionStrategy.CountExistsShortCircuit;
    }
}

File: src/java/foundgine-core/src/main/java/com/foundgine/core/semantic/planning/AggregateRelationshipFilterPushdownRule.java
package com.foundgine.core.semantic.planning;

import com.foundgine.core.semantic.query.*;
import java.util.*;

/**
 * Pushes an existential relationship predicate into a COUNT aggregate. COUNT(R)
 * > 0 AND SOME(R, P) becomes COUNT(R WHERE P) > 0.
 */
public final class AggregateRelationshipFilterPushdownRule implements IPlanRewriteRule {
	@Override
	public String name() {
		return "aggregate.relationship-filter.pushdown";
	}

	@Override
	public List<String> preconditions() {
		return List.of("plan contains an AND filter", "AND contains an eligible COUNT existence predicate",
				"AND contains a SOME relationship predicate for the same relationship",
				"COUNT aggregate does not already contain a predicate");
	}

	@Override
	public List<String> securityObligations() {
		return List.of("authorization.required", "authorization.runtime", "visibility.relationship",
				"planning.cache-context-isolation");
	}

	@Override
	public double costImpact() {
		return 1.25d;
	}

	@Override
	public double benefitEstimate() {
		return 4d;
	}

	@Override
	public List<String> mustRunBefore() {
		return List.of("aggregate.cardinality.short-circuit");
	}

	@Override
	public int priority() {
		return 35;
	}

	@Override
	public boolean canApply(SemanticPlan plan) {
		Objects.requireNonNull(plan);
		return containsEligible(plan.root());
	}

	@Override
	public SemanticPlan apply(SemanticPlan plan) {
		Objects.requireNonNull(plan);
		if (!canApply(plan))
			return plan;
		boolean[] changed = { false };
		var root = rewriteNode(plan.root(), changed);
		return changed[0] ? new SemanticPlan(root, plan.requiredSecurityInvariants(), plan.authorizationBinding())
				: plan;
	}

	private static SemanticPlanNode rewriteNode(SemanticPlanNode node, boolean[] changed) {
		var options = node.queryOptions();
		boolean filterChanged = false;
		if (options != null && options.filter() != null) {
			var original = options.filter();
			var rewritten = rewriteFilter(original, changed);
			filterChanged = rewritten != original;
			if (filterChanged)
				options = new SemanticQueryOptions(rewritten, options.order(), options.limit(), options.offset(),
						options.after());
		}
		List<SemanticPlanNode> children = new ArrayList<>(node.children().size());
		boolean childChanged = false;
		for (var child : node.children()) {
			var r = rewriteNode(child, changed);
			children.add(r);
			childChanged |= r != child;
		}
		if (childChanged)
			changed[0] = true;
		if (filterChanged)
			changed[0] = true;
		if (filterChanged || childChanged)
			return new SemanticPlanNode(node.id(), node.operation(), node.entityId(), node.fields(),
					node.viaRelationship(), node.viaConnection(), children, options, node.authorization(),
					node.relationshipCardinality(), node.traversalMode(), node.traversalOrder(),
					filterChanged ? AggregateExecutionStrategy.DEFAULT : node.aggregateExecutionStrategy());
		return node;
	}

	private static SemanticFilterExpression rewriteFilter(SemanticFilterExpression filter, boolean[] changed) {
		if (filter instanceof SemanticAndFilter and) {
			var expressions = new ArrayList<>(and.expressions());
			for (int i = 0; i < expressions.size(); i++) {
				var aggregate = expressions.get(i);
				if (!(aggregate instanceof SemanticAggregateFilter a) || !isEligibleCountExists(a))
					continue;
				for (int j = 0; j < expressions.size(); j++) {
					if (i == j || !(expressions.get(j) instanceof SemanticRelationshipFilter r)
							|| r.quantifier() != SemanticRelationshipQuantifier.SOME
							|| !r.relationship().equals(a.relationship()))
						continue;
					var rewritten = new SemanticAggregateFilter(a.relationship(), a.aggregate(), a.field(),
							a.operator(), a.value(), r.predicate());
					var remaining = new ArrayList<SemanticFilterExpression>();
					for (int k = 0; k < expressions.size(); k++) {
						if (k == i)
							remaining.add(rewritten);
						else if (k != j)
							remaining.add(expressions.get(k));
					}
					changed[0] = true;
					if (remaining.isEmpty())
						throw new IllegalStateException("Aggregate pushdown produced an empty AND expression.");
					return remaining.size() == 1 ? remaining.get(0) : new SemanticAndFilter(remaining);
				}
			}
			var nested = new ArrayList<SemanticFilterExpression>(expressions.size());
			boolean nestedChanged = false;
			for (var expression : expressions) {
				var r = rewriteFilter(expression, changed);
				nested.add(r);
				nestedChanged |= r != expression;
			}
			return nestedChanged ? new SemanticAndFilter(nested) : filter;
		}
		if (filter instanceof SemanticOrFilter or) {
			var nested = new ArrayList<SemanticFilterExpression>(or.expressions().size());
			boolean nestedChanged = false;
			for (var expression : or.expressions()) {
				var r = rewriteFilter(expression, changed);
				nested.add(r);
				nestedChanged |= r != expression;
			}
			return nestedChanged ? new SemanticOrFilter(nested) : filter;
		}
		if (filter instanceof SemanticRelationshipFilter relationship) {
			var predicate = rewriteFilter(relationship.predicate(), changed);
			return predicate == relationship.predicate() ? relationship
					: new SemanticRelationshipFilter(relationship.relationship(), relationship.quantifier(), predicate);
		}
		return filter;
	}

	private static boolean containsEligible(SemanticPlanNode node) {
		return (node.queryOptions() != null && node.queryOptions().filter() != null
				&& containsEligible(node.queryOptions().filter()))
				|| node.children().stream().anyMatch(AggregateRelationshipFilterPushdownRule::containsEligible);
	}

	private static boolean containsEligible(SemanticFilterExpression filter) {
		if (filter instanceof SemanticAndFilter and)
			return and.expressions().stream()
					.anyMatch(x -> x instanceof SemanticAggregateFilter a && isEligibleCountExists(a))
					&& and.expressions().stream().anyMatch(x -> x instanceof SemanticRelationshipFilter r
							&& r.quantifier() == SemanticRelationshipQuantifier.SOME);
		if (filter instanceof SemanticRelationshipFilter r)
			return containsEligible(r.predicate());
		if (filter instanceof SemanticOrFilter or)
			return or.expressions().stream().anyMatch(AggregateRelationshipFilterPushdownRule::containsEligible);
		return false;
	}

	private static boolean isEligibleCountExists(SemanticAggregateFilter a) {
		return a.aggregate() == SemanticFilterAggregate.COUNT && a.field() == null && a.predicate() == null
				&& AggregateExecutionStrategyResolver.resolve(a.operator(),
						a.value()) == AggregateExecutionStrategy.COUNT_EXISTS_SHORT_CIRCUIT;
	}
}
